[export] Custom object serialization

angelayi opened the pull request:
Some NvidaTRT folks were asking for a way to integrate the serialization of custom objects with export's serialization. After some discussion (more background [here](https://docs.google.com/document/d/1lJfxakmgeoEt50inWZ53MdUtOSa_0ihwCuPy_Ak--wc/edit)), we settled on a way for users to register their custom object's serializer/deserializer functions.

Since TorchScript's `.def_pickle` already exists for [registering custom classes](https://pytorch.org/tutorials/advanced/torch_script_custom_classes.html), and `tensorrt.ICudaEngine` already contains a `.def_pickle` implementation, we'll start off by reusing the existing framework and integrating it with export's serialization. 

TorchScript's `.def_pickle` requires users to register two functions, which end up being the `__getstate__` and `__setstate__` methods on the class. The semantics of `__getstate__` and `__setstate__` in TorchScript are equivalent to that of Python pickle modules. This is then registered using pybind's `py::pickle` function [here](https://www.internalfb.com/code/fbsource/[f44e048145e4697bccfaec300798fce7daefb858]/fbcode/caffe2/torch/csrc/jit/python/script_init.cpp?lines=861-916) to be used with Python's pickle to initialize a ScriptObject with the original class, and set the state back to what it used to be.

I attempted to call `__getstate__` and `__setstate__` directly, but I couldn't figure out how to initial the object to be called with `__setstate__` in python. One option would be to create a `torch._C.ScriptObject` and then set the class and call `__setstate__`, but there is no constructor initialized for ScriptObjects. Another option would be to construct an instance of the serialized class itself, but if the class constructor required arguments, I wouldn't know what to initialize it with. In ScriptObject's `py::pickle` registration it directly creates the object [here](https://www.internalfb.com/code/fbsource/[f44e048145e4697bccfaec300798fce7daefb858]/fbcode/caffe2/torch/csrc/jit/python/script_init.cpp?lines=892-906), which is why I was thinking that just directly using Python's `pickle` will be ok since it is handled here.

So, what I did is that I check if the object is pickle-able, meaning it contains `__getstate__` and `__setstate__` methods, and if so, I serialize it with Python's pickle. TorchScript does have its own implementation of [pickle/unpickle](https://www.internalfb.com/code/fbsource/[59cbc569ccbcaae0db9ae100c96cf0bae701be9a][history]/fbcode/caffe2/torch/csrc/jit/serialization/pickle.h?lines=19%2C82), but it doesn't seem to have pybinded functions callable from python.

A question is -- is it ok to combine this pickle + json serialization?

angelayi committed:
[export] Custom object serialization

vadimkantorov commented:
I don't know if it's related, but in many contexts it's important to not do any arbitrary unpickling on load by default for security reasons:
- https://github.com/pytorch/pytorch/issues/52181

tugsbayasgalan commented on torch/_export/serde/serialize.py:
```diff
@@ -582,6 +585,14 @@ def serialize_optional_tensor_args(a):
             return Argument.create(as_memory_format=_TORCH_TO_SERIALIZE_MEMORY_FORMAT[arg])
         elif isinstance(arg, torch.layout):
             return Argument.create(as_layout=_TORCH_TO_SERIALIZE_LAYOUT[arg])
+        elif hasattr(type(arg), "__getstate__") and hasattr(type(arg), "__setstate__"):
```
Should we check here that arg is indeed torch.classes? IIUC, this will only work for Torchbind classes right?

tugsbayasgalan commented on test/export/test_serialize.py:
```diff
@@ -416,5 +423,49 @@ def test_model_op_namespace_version_missing_from_deserializer_do_not_raises(self
+            if node.op == "call_function" and node.target == torch.ops.aten.add.Tensor:
+                arg0, arg1 = node.args
+                self.assertTrue(isinstance(arg1, torch._C.ScriptObject))
+                self.assertEqual(arg1.__getstate__(), custom_obj.__getstate__())
```
NIT: Maybe also check that arg1 can do all Torchbind things like calling method etc on the custom type?

angelayi committed:
comments

gmagogsfm commented on test/export/test_serialize.py:
```diff
@@ -416,5 +423,50 @@ def test_model_op_namespace_version_missing_from_deserializer_do_not_raises(self
+                arg0, arg1 = node.args
+                node.args = (arg0, custom_obj)
+
+        # Since the exported program is not runnable with our custom obj, we
```
Per offline discussion, let's use an op that accepts the custom class object to fully test deserializer logic

angelayi committed:
custom op

angelayi commented:
Updated the test case with a custom operator using the custom object. This required catching an exception made during torch.fx.graph_module's `recompile` function. This is because with the custom object, the graph looks like the following:
```
graph():
    %arg0_1 : [num_users=1] = placeholder[target=arg0_1]
    %take_an_instance_default : [num_users=1] = call_function[target=torch.ops._TorchScriptTesting.take_an_instance.default](args = (ScriptObject <__torch__.torch.classes._TorchScriptTesting._PickleTester>,), kwargs = {})
    %add : [num_users=1] = call_function[target=torch.ops.aten.add.Tensor](args = (%arg0_1, %take_an_instance_default), kwargs = {})
    return (add,)
```
Resulting in the generated python looking like:
```
def forward(self, arg0_1):
    take_an_instance_default = torch.ops._TorchScriptTesting.take_an_instance.default(<torch.ScriptObject object at 0x562b3191f620>)
    add = torch.ops.aten.add.Tensor(arg0_1, take_an_instance_default);  arg0_1 = take_an_instance_default = None
    return (add,)
```
However, this is not runnable as python is unable to recognize what the string `<torch.ScriptObject object at 0x562b3191f620>` represents. When running the graph using `torch.fx.Interpreter`, which is what ExportedProgram does, this works as it is able to map `<torch.ScriptObject object at 0x562b3191f620>` to the actual object in memory. So I just bypassed the `recompile()` error and added a warning.

gmagogsfm commented:
> Updated the test case with a custom operator using the custom object. This required catching an exception made during torch.fx.graph_module's `recompile` function. This is because with the custom object, the graph looks like the following:
> 
> ```
> graph():
>     %arg0_1 : [num_users=1] = placeholder[target=arg0_1]
>     %take_an_instance_default : [num_users=1] = call_function[target=torch.ops._TorchScriptTesting.take_an_instance.default](args = (ScriptObject <__torch__.torch.classes._TorchScriptTesting._PickleTester>,), kwargs = {})
>     %add : [num_users=1] = call_function[target=torch.ops.aten.add.Tensor](args = (%arg0_1, %take_an_instance_default), kwargs = {})
>     return (add,)
> ```
> 
> Resulting in the generated python looking like:
> 
> ```
> def forward(self, arg0_1):
>     take_an_instance_default = torch.ops._TorchScriptTesting.take_an_instance.default(<torch.ScriptObject object at 0x562b3191f620>)
>     add = torch.ops.aten.add.Tensor(arg0_1, take_an_instance_default);  arg0_1 = take_an_instance_default = None
>     return (add,)
> ```
> 
> However, this is not runnable as python is unable to recognize what the string `<torch.ScriptObject object at 0x562b3191f620>` represents. When running the graph using `torch.fx.Interpreter`, which is what ExportedProgram does, this works as it is able to map `<torch.ScriptObject object at 0x562b3191f620>` to the actual object in memory. So I just bypassed the `recompile()` error and added a warning.

You mean generated python code is not runnable? That's kinda expected. But the GraphModule and ExportedProgram should be runnable I presume?

angelayi commented:
> You mean generated python code is not runnable? That's kinda expected. But the GraphModule and ExportedProgram should be runnable I presume?

The generated python code is not runnable, making the GraphModule not runnable if you just directly call it eagerly. But you can still run it through torch.fx.Interpreter. ExportedProgram is still runnable because we run the graph module using torch.fx.Interpreter

angelayi committed:
pickleTester is dumb

angelayi commented:
@pytorchbot merge
